src/Data/List/NonEmpty.purs

```diff
@@ -0,0 +1,84 @@
+module Data.List.NonEmpty where
+
+import Prelude
+
+import Control.Alt (class Alt)
+
+import Data.Foldable (class Foldable)
+import Data.Generic (class Generic)
+import Data.List ((:))
+import Data.List as L
+import Data.Newtype (class Newtype, unwrap)
+import Data.NonEmpty (NonEmpty, (:|))
+import Data.NonEmpty as NE
+import Data.Maybe (fromMaybe)
+import Data.Traversable (class Traversable)
+import Data.Tuple (Tuple(..))
+import Data.Unfoldable (class Unfoldable, unfoldr)
+
+newtype NonEmptyList a = NonEmptyList (NonEmpty L.List a)
+
+toUnfoldable :: forall f a. Unfoldable f => NonEmptyList a -> f a
+toUnfoldable (NonEmptyList nel) =
+  unfoldr (\xs -> (\rec -> Tuple rec.head rec.tail) <$> L.uncons xs) (NE.oneOf nel)
+
+singleton :: forall a. a -> NonEmptyList a
+singleton = NonEmptyList <<< NE.singleton
+
+head :: forall a. NonEmptyList a -> a
+head (NonEmptyList (x :| _)) = x
+
+last :: forall a. NonEmptyList a -> a
+last (NonEmptyList (x :| xs)) = fromMaybe x (L.last xs)
+
+tail :: forall a. NonEmptyList a -> L.List a
+tail (NonEmptyList (_ :| xs)) = xs
+
+init :: forall a. NonEmptyList a -> NonEmptyList a
+init (NonEmptyList (x :| xs)) = NonEmptyList (x :| fromMaybe L.Nil (L.init xs))
+
+uncons :: forall a. NonEmptyList a -> { head :: a, tail :: L.List a }
+uncons (NonEmptyList (x :| xs)) = { head: x, tail: xs }
+
+length :: forall a. NonEmptyList a -> Int
+length (NonEmptyList (x :| xs)) = 1 + L.length xs
+
+concatMap :: forall a b. (a -> NonEmptyList b) -> NonEmptyList a -> NonEmptyList b
+concatMap f (NonEmptyList (a :| as)) =
+  case f a of
+    NonEmptyList (b :| bs) ->
+      NonEmptyList (b :| bs <> L.concatMap (NE.oneOf <<< unwrap <<< f) as)

```

Comment:
Good idea. I guess I was trying to resist that since we dropped all the specialised `toLists` in favour of `fromFoldable`, but it makes sense here I think!